File: template_searching.py
# BINARY SEARCH FUNCTIONS

# returns -1 if no value satisfies

def find_smallest_true(lo, hi, check):
    left, right = lo, hi
    ans = -1
    while left <= right:
        mid = (left + right) // 2
        if check(mid):
            ans = mid
            right = mid - 1
        else:
            left = mid + 1
    return ans

def find_largest_true(lo, hi, check):
    left, right = lo, hi
    ans = -1
    while left <= right:
        mid = (left + right) // 2
        if check(mid):
            ans = mid
            left = mid + 1
        else:
            right = mid - 1
    return ans


# TERNARY SEARCH FOR MINIMUM (INTEGER VERSION)

def ternary_search(lo: int, hi: int, check) -> int:
    while hi - lo > 3:
        m1 = lo + (hi - lo) // 3
        m2 = hi - (hi - lo) // 3
        if check(m1) < check(m2):
            hi = m2 - 1
        else:
            lo = m1 + 1
    best = lo
    for x in range(lo, hi + 1):
        if check(x) < check(best):
            best = x

def ternary_search(f, left, right, eps=1e-7):
    """
    Find x in [left, right] that minimizes f(x).
    f must be unimodal (convex in this case).
    eps = precision
    """
    while right - left > eps:
        m1 = left + (right - left) / 3
        m2 = right - (right - left) / 3
        
        if f(m1) < f(m2):
            right = m2   # min is in [left, m2]
        else:
            left = m1    # min is in [m1, right]
    
    return (left + right) / 2  # approximate min point

File: cpp-templates/template_searching.cpp
#include <bits/stdc++.h>
using namespace std;

template <class Check>
long long find_smallest_true(long long lo, long long hi, Check check) {
    long long ans = -1;
    while (lo <= hi) {
        long long mid = lo + (hi - lo) / 2;
        if (check(mid)) {
            ans = mid;
            hi = mid - 1;
        } else {
            lo = mid + 1;
        }
    }
    return ans;
}

template <class Check>
long long find_largest_true(long long lo, long long hi, Check check) {
    long long ans = -1;
    while (lo <= hi) {
        long long mid = lo + (hi - lo) / 2;
        if (check(mid)) {
            ans = mid;
            lo = mid + 1;
        } else {
            hi = mid - 1;
        }
    }
    return ans;
}

template <class F>
long long ternary_search_int(long long lo, long long hi, F f) {
    while (hi - lo > 3) {
        long long m1 = lo + (hi - lo) / 3;
        long long m2 = hi - (hi - lo) / 3;
        if (f(m1) < f(m2)) hi = m2 - 1;
        else lo = m1 + 1;
    }
    long long best = lo;
    for (long long x = lo; x <= hi; x++) {
        if (f(x) < f(best)) best = x;
    }
    return best;
}

template <class F>
double ternary_search_double(double left, double right, F f, double eps = 1e-7) {
    while (right - left > eps) {
        double m1 = left + (right - left) / 3.0;
        double m2 = right - (right - left) / 3.0;
        if (f(m1) < f(m2)) right = m2;
        else left = m1;
    }
    return (left + right) / 2.0;
}
